src/librustc_codegen_ironox/builder.rs

```diff
@@ -0,0 +1,940 @@
+    type Value = <CodegenCx<'ll, 'tcx> as BackendTypes>::Value;
+    type BasicBlock = <CodegenCx<'ll, 'tcx> as BackendTypes>::BasicBlock;
+    type Type = <CodegenCx<'ll, 'tcx> as BackendTypes>::Type;
+    type Context = <CodegenCx<'ll, 'tcx> as BackendTypes>::Context;
+    type Funclet = <CodegenCx<'ll, 'tcx> as BackendTypes>::Funclet;
+
+    type DIScope = <CodegenCx<'ll, 'tcx> as BackendTypes>::DIScope;
+}
+
+pub struct Builder<'a, 'll: 'a, 'tcx: 'll> {
+    pub cx: &'a CodegenCx<'ll, 'tcx>,
+}
+
+impl ty::layout::LayoutOf for Builder<'_, '_, 'tcx> {
+    type Ty = Ty<'tcx>;
+    type TyLayout = TyLayout<'tcx>;
+
+    fn layout_of(&self, ty: Ty<'tcx>) -> Self::TyLayout {
+        self.cx.layout_of(ty)
+    }
+}
+
+impl ty::layout::HasDataLayout for Builder<'_, '_, '_> {
+    fn data_layout(&self) -> &ty::layout::TargetDataLayout {
+        self.cx.data_layout()
+    }
+}
+
+impl ty::layout::HasTyCtxt<'tcx> for Builder<'_, '_, 'tcx> {
+    fn tcx<'a>(&'a self) -> TyCtxt<'a, 'tcx, 'tcx> {
+        self.cx.tcx
+    }
+}
+
+impl HasCodegen<'tcx> for Builder<'a, 'll, 'tcx> {
+    type CodegenCx = CodegenCx<'ll, 'tcx>;
+}
+
+impl BuilderMethods<'a, 'tcx> for Builder<'a, 'll, 'tcx> {
+    fn new_block<'b>(
+        cx: &'a Self::CodegenCx,
+        llfn: <Self::CodegenCx as BackendTypes>::Value,
+        name: &'b str
+    )-> Self {
+        let bx = Builder::with_cx(cx);
+        // add a basic block
+        bx
+    }
+
+    fn with_cx(cx: &'a Self::CodegenCx) -> Self {
+        Builder {
+            cx,
+        }
+    }
+
+    fn build_sibling_block<'b>(&self, name: &'b str) -> Self {
+        Builder::new_block(self.cx, self.llfn(), name)
+    }
+    fn cx(&self) -> &CodegenCx<'ll, 'tcx> {
+        &self.cx
+    }
+    fn llfn(&self) -> &'ll Value {
+        &Value {}
+    }
+    fn llbb(&self) -> &'ll BasicBlock {
+        &BasicBlock {}
+    }
+    fn count_insn(&self, category: &str) {
+        unimplemented!("count_insn(&self, category: &str)");
+    }
+
+    fn set_value_name(&mut self, value: <Self::CodegenCx as BackendTypes>::Value, name: &str) {
+        unimplemented!("set_value_name");
+    }
+    fn position_at_end(&mut self, llbb: <Self::CodegenCx as BackendTypes>::BasicBlock) {
+        unimplemented!("position_at_end");
+    }
+    fn position_at_start(&mut self, llbb: <Self::CodegenCx as BackendTypes>::BasicBlock) {
+        unimplemented!("position_at_start");
+    }
+    fn ret_void(&mut self) {
+        unimplemented!("ret_void");
+    }
+    fn ret(&mut self, v: <Self::CodegenCx as BackendTypes>::Value) {
+        unimplemented!("ret");
+    }
+    fn br(&mut self, dest: <Self::CodegenCx as BackendTypes>::BasicBlock) {
+        unimplemented!("br");
+    }
+    fn cond_br(
+        &mut self,
+        cond: <Self::CodegenCx as BackendTypes>::Value,
+        then_llbb: <Self::CodegenCx as BackendTypes>::BasicBlock,
+        else_llbb: <Self::CodegenCx as BackendTypes>::BasicBlock,
+    ) {
+        unimplemented!("");
```

Comment:
Is it worth populating the error strings like you did elsewhere?

Reply:
Will do. I was a bit lazy: I copied the trait definition and only wrote a regex for the simple signatures.